Ledger query simplification

clemahieu opened the pull request:
This PR merges the ledger::account/balance/amount variants in to a single function for each, returning an optional associated value.

The major advantage is a simplification in the nano::ledger interface and increased type-checking in values returned.  Previously there were separate functions for assumed-correct function calls protected with debug_assert, and variants permitted to fail that returned error codes or sentinel values.

These changes replace each account/amount/balance with functions that return std::optional values for each associated type. Variants that used assumed-correct functions simply call std::optional::value, which does runtime value checking which increases correctness safety. Variants that used _safe variants or checked sentinel values are replaced with std::optional bool conversion checks and operate the same way.

clemahieu committed:
Make use of ledger::account overload taking a block when block is already available.

clemahieu committed:
Merging ledger::balance_safe with ledger::balance and returning value as an std::optional.

clemahieu committed:
Merge ledger::amount_safe in to ledger::amount and change return value to std::optional

clemahieu committed:
Merge the ledger::account and ledger::account_safe functions in to a single function returning an optional value.

This simplifies the ledger interface and uses safer std::optional to signal no value rather than a sentinel 0 value.

pwojcikdev commented on nano/nano_node/entry.cpp:
```diff
@@ -1494,7 +1494,10 @@ int main (int argc, char * const * argv)
+								if (auto balance = node->ledger.balance (transaction, state_block.hashables.previous))
+								{
+									prev_balance = balance.value ();
+								}
```
Those `if (opt = ...) { value = opt.value () }` blocks could in many cases be replaced simply by one-liner: `opt.value_or({})` but it's a cosmetic improvement - new style is better either way.

clemahieu replied:
That's a good point. I'll convert these as I like that style better.

pwojcikdev commented on nano/node/bootstrap_ascending/service.cpp:
```diff
@@ -132,7 +132,7 @@ void nano::bootstrap_ascending::service::inspect (store::transaction const & tx,
 		case nano::block_status::progress:
 		{
-			const auto account = ledger.account (tx, hash);
+			const auto account = ledger.account (block);
```
Not reusing existing db transaction is a big red flag, given that this is called very frequently, what's the performance impact here?

pwojcikdev commented on nano/node/json_handler.cpp:
```diff
@@ -1150,13 +1150,12 @@ void nano::json_handler::block_info ()
+				response_l.put ("amount", amount.value ().convert_to<std::string> ());
 			}
-			auto balance (node.ledger.balance (transaction, hash));
+			auto balance = node.ledger.balance (*block);
```
Not reusing existing db transaction.

pwojcikdev commented on nano/node/json_handler.cpp:
```diff
@@ -1214,20 +1213,19 @@ void nano::json_handler::block_confirm ()
-				auto account (node.ledger.account (transaction, hash));
-				bool error_or_pruned (false);
-				auto amount (node.ledger.amount_safe (transaction, hash, error_or_pruned));
+				auto account = node.ledger.account (*block_l);
```
Not reusing existing db transaction - there are many call sites like this. Is there any reason for those changes? Reusing db transactions seems to me like a simple and positive thing, it's both faster and guarantees consistent view - something that might be important down the road with database sharding.

clemahieu replied:
These conversions from (tx, hash) to (block &) are misleading, they don't open new transactions on the database, these are utility functions that are hung on the ledger class that pull the values from the block data or sideband.

I'm going to make a followup PR that moves these on to the block itself.

pwojcikdev replied:
Ahh, I see, it now makes sense. I should have started reviewing this PR by looking at the ledger class itself.

pwojcikdev commented on nano/secure/ledger.cpp:
```diff
@@ -834,30 +831,18 @@ nano::uint128_t nano::ledger::balance (nano::block const & block)
 {
 	if (hash.is_zero ())
 	{
 		return 0;
```
Being consistent with the new style of using optionals to indicate failed ops, should this perhaps return `std::nullopt` ?

clemahieu replied:
Made this change. It was a bit more tricky because ledger::amount was depending on this returning 0 if it couldn't find the hash.  I added an extra commit to change ledger::amount so it no longer depends on this behaviour.

pwojcikdev commented on nano/secure/ledger.cpp:
```diff
@@ -1156,68 +1141,32 @@ nano::account nano::ledger::account (nano::block const & block) const
+std::optional<nano::uint128_t> nano::ledger::amount (store::transaction const & transaction_a, nano::block_hash const & hash_a)
 {
 	auto block_l = block (transaction_a, hash_a);
 	debug_assert (block_l);
```
Shouldn't we check if block is empty and return nullopt instead of keeping this debug assert?

clemahieu committed:
ledger::amount will no longer query for the previous block's balance if the previous hash is zero.

This removes a dependency on the behaviour of ledger::balance to return a sentinel value of 0 if the block hash cannot be found.

clemahieu commented on nano/node/bootstrap/bootstrap_server.cpp:
```diff
@@ -286,7 +286,10 @@ nano::asc_pull_ack nano::bootstrap_server::process (const store::transaction & t
 		{
 			// Try to lookup account assuming target is block hash
-			target = ledger.account_safe (transaction, request.target.as_block_hash ());
+			if (auto account = ledger.account (transaction, request.target.as_block_hash ()))
```
Missed one value_or
